src/wasm/wasm.cpp

```diff
@@ -362,41 +301,55 @@ static void handleUnreachable(Block* block,
 }
 
 void Block::finalize() {
+  if (list.size() == 0) {
+    type = Type::none;
+    return;
+  }
   if (!name.is()) {
-    if (list.size() > 0) {
-      // nothing branches here, so this is easy
-      // normally the type is the type of the final child
-      type = list.back()->type;
-      // and even if we have an unreachable child somewhere,
-      // we still mark ourselves as having that type,
-      // (block (result i32)
-      //  (return)
-      //  (i32.const 10)
-      // )
-      if (type.isConcrete()) {
-        return;
-      }
-      // if we are unreachable, we are done
-      if (type == Type::unreachable) {
+    // nothing branches here, so this is easy
+    // normally the type is the type of the final child
+    type = list.back()->type;
+    // and even if we have an unreachable child somewhere,
+    // we still mark ourselves as having that type,
+    // (block (result i32)
+    //  (return)
+    //  (i32.const 10)
+    // )
+    if (type.isConcrete()) {
+      return;
+    }
+    // if we are unreachable, we are done
+    if (type == Type::unreachable) {
+      return;
+    }
+    // we may still be unreachable if we have an unreachable
+    // child
+    for (auto* child : list) {
+      if (child->type == Type::unreachable) {
+        type = Type::unreachable;
         return;
       }
-      // we may still be unreachable if we have an unreachable
-      // child
-      for (auto* child : list) {
-        if (child->type == Type::unreachable) {
-          type = Type::unreachable;
-          return;
-        }
-      }
-    } else {
-      type = Type::none;
     }
     return;
   }
 
-  TypeSeeker seeker(this, this->name);
-  type = Type::mergeTypes(seeker.types);
-  handleUnreachable(this);
+  // The default type is according to the value that flows out.
+  type = list.back()->type;
+  BranchUtils::BranchSeeker seeker(this->name);
+  Expression* temp = this;
+  seeker.walk(temp);
```

Comment:
Would `seeker.walk(this)` work?

Reply:
No, as that function takes a reference, which it can potentially modify (all walkers can do so in general, if you call `replaceCurrent`).

It would be nice if we could have two walkers,
```cpp
walk(Expression*&)
walk(Expression*)
```
and the latter would be a fallback for a case like this. But C++ doesn't seem to allow that...

Reply:
Maybe we could give `Walker` a helper method like `walkImmutable` that takes care of this temporary variable dance, but I don't love that solution either.